servant-client-core/src/Servant/Client/Core/Internal/RunClient.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeFamilies          #-}
+-- | Types for possible backends to run client-side `Request` queries
+module Servant.Client.Core.Internal.RunClient where
+
+import           Prelude.Compat
+import           Prelude                              ()
+
+import           Control.Monad                        (unless)
+import           Control.Monad.Error.Class            (MonadError, throwError)
+import           Data.Foldable                        (toList)
+import           Data.Proxy                           (Proxy)
+import qualified Data.Text                            as T
+import           Network.HTTP.Media                   (MediaType, matches,
+                                                       parseAccept, (//))
+import           Servant.API                          (MimeUnrender,
+                                                       contentTypes,
+                                                       mimeUnrender)
+import           Servant.Client.Core.Internal.Request (Request, Response (..),
+                                                       ServantError (..))
+
+class (MonadError ServantError m) => RunClient m where
```

Comment:
I'd make a `class AsServantError e where buildServantError :: ServantError -> e`, or actual `Prism`

Reply:
And then `(MonadError e m, AsServantError e)`?

Reply:
Ah, you need to catch errors too. Then one would need `Prism`.

It would be nice to have `(MonadError e m, AsServantError e)` as then one doesn't really need `enter` in client side, as one could make its `MegaMonad` an instance of `RunClient`. But to be composable we need `AsServantError` (as `MegaMonad` might want to throw and catch other errors).

I'm completely ok to depend on `lens` (or only `profunctors` to get `Prism`s). But we can do that change later.

Reply:
The other option is to require `MonadCatch e` instead of `(MonadError e m, AsServantError e)`.

Reply:
```haskell
class RunClient m where
  runRequest :: Request -> m Response
  throwServantError :: ServantError -> m ()
  catchServantError :: m a -> (ServantError -> m a) -> m a
```

`IO` can be instance of this, but we still tell what we `throw`/`catch`

Reply:
:+1: - done!